---
title: "Seurat and R Coding Training"
author: "Ryan"
date: "4/10/2020"
output: html_document
---

##Load Packages

```{r load.packages}
library(Seurat)
library(ggplot2)
library(dplyr)
library(patchwork)
```

##Load Seurat Objects

```{r load.object}
load("training.object.Rda")
```

# Seurat Object Meta Data

In this part of the tutorial, we are reviewing how to access information in the object@meta.data slot of a Seurat object and create new meta.data columns for cells. Meta data (or cell-specific characteristics) are stored in the Seurat object under the object@meta.data slot. These cell-specific characteristics can include information like patient ID, number of genes expressed by the cell, or sequencing technology. The default information that is stored in the object@meta.data slot when the Seurat Object is first created is:

nCount_RNA: number of transcripts expressed by the cell (related to sequencing reads per cell)
nFeature_RNA: number of genes expressed >0 by the cell (remember that when you create a Seurat object, you specify to only include genes that are exprssed >0 in X cells)
orig.ident: this column is automatically filled in by whatever variable is present in the project="X" part of CreateSeuratObject(project = "X")

All other meta.data about the cell that is known a priori will need to be added manually. We will go over how to do that in the following sections.

First, let's take a look at what the object@meta.object actually looks like. For training purposes, our Seurat object is called training.object, but you will need to change that to whatever is the name of the Seurat object you are analyzing.

Execute the following code:

```{r exploring.meta.data}
training.object@meta.data[1:5,1:5]
```

The output of this shows us the first 5 rows and first 5 columns of our meta data. "@" accesses slots in the Seurat object. To get information about cells, you want to access the @meta.data slot. Notice that the @meta.data slot is arranged such that all of the rows represent the cell ID and the columns represent different information about the cells.

Your Seurat object can store unlimited types of cell-specific information. When you are first given a Seurat object that has already been created, the first step is to learn what type of information has already been stored in your Seurat object.

In order to learn what types of information about your cells is stored in your Seurat Object you execute the following code: 

```{r accessing.meta.data}
colnames(training.object@meta.data)
```

Notice the output of this command lists several variables, all containing different information about the cells in our dataset. Let's explore some of them now.

If you are unsure what information is stored in each meta.data column, it is helpful to identify what are the variables contained in that column. For training purposes, let's find out what is stored in the "tissue" column of our meta data.

In order to learn what types of information about your cells is stored in a specific column of your meta data you execute the following code: 

```{r unique.meta.data.column}
unique(training.object@meta.data$tissue)
```

"$" accesses specific columns in the @meta.data slot.

Notice the output of this lists two variables "Lung" and "Tumor". This indicates that every cell in that column of the dataset has been assigned a value of either "Lung" or "Tumor", indicating that this column identifies the tissue of origin that these cells were isolated from. Try changing the "tissue" variable to "paper.ident" or "Phase" to learn what other information about the cells is stored.

Sometimes, you will want to add new information about a cell that is not already contained in the meta data of a seurat object. This is most often accomplished in one of two scenarios:

1) You know the specific cell IDs (i.e. the rownames of our  @meta.data slot) that neeed to have new meta data added to them
2) You want to add new meta data to a cell based on information already present in the @meta.data slot

# Adding New Meta Data from Known Cell IDs

Lets go over the first scenario now. We have been given new information about the cells in our dataset that allows us to know which cells are doublets and singlets. This information was provided to us as a list of cell names. Let's go ahead and load  that list and review it. For training purposes, this list is called doublet_ids.txt, but you will need to change it depending on your circumstances.

To load the list of cell IDs for the doublets execute the following code:

```{r read.in.doublet.ids}
doublets.ids<-read.table("doublet_ids.txt")
doublets.ids
```

Notice the output of this code is a matrix with one column and 10 rows. Each row is the identity of a cell in our dataset that is known to be a doublet. In order to search our dataset for these cells, we will first need to convert this matrix into a string of characters. 

To convert the matrix of doublet IDs  to a list of characters execute the following code:

```{r convert.doublets.to.characters}
doublets.ids<-as.character(doublets.ids$x)
doublets.ids
```

We now have the list of cell IDs that are known to be doublets. We were given 10 cell IDs, but in order to add new meta data columns, you need to add information for every cell. Let's see how many cells exist in our dataset. To do this, we will use rownames() function.

To determine how many cells exist in our dataset execute the following code:

```{r how.many.cells}
rownames(training.object@meta.data)
```

The output of this lists all the cells present in our dataset, which we can see is 247. However, for datasets of >10000 cells, this is not a very effective way to count the cells in our dataset. Instead, we can use the length() function to output the length of the rownames() function:

```{r length.of.cell.in.dataset}
length(rownames(training.object@meta.data))
```

Same information, but easier to interpret. Back to our original issue: we have been given 10 cell IDs that represent doublets, but no information on the other 237 cells. When we create our new meta.data, we will have to remember to add information for all the cells.

Seurat accepts new information to be added to the metadata in the form of a series of characters, where each character corresponds to information about a cell. It is important that the order of these characters exactly matches the order of the cells in the meta data of our seurat object. In order to create this list, we will be using a combination of the list(), which(), and as.character() functions.

The first step of this process is to create an empty list that we can fill in with our new meta data. To do that , execute the following code:

```{r create.empty.meta.data.doublets}
doublets.metadata<-list()
doublets.metadata
```

The output of this command is an empty list. We will want to fill in this list with our new meta data. We will start by calling every cell as a "singlet", then rewrite those cells that we know to be doublets later. In order to set a component of a list we will use the following command:

```{r add.singlet.to.list}
doublets.metadata[1]<-"singlet"
doublets.metadata
```

Notice, that this command only added one "singlet" character to our list. However, we will need to add information about all 247 cells. This can be done easily by specifying additional components of our list:

```{r add.multiple.singlet}
doublets.metadata[1:247]<-"singlet"
head(doublets.metadata)
```

You can see that the "singlet" character has been added to all 247 components in our list. The head() command will only print the first few components of any variable you provide. You can confirm that all 247 components of the list are called "singlet" by removing the head() function if you prefer. 

Now that we generated information about all 247 cells, we will need to specify which 10 cells are the doublets. In order to do this, we will need to find the rownames of our @meta.data slot which match the cell IDs from our doublet.ids variable. To do this we will use the which() function.

The which() function is part of base R and allows you to pattern match two different variables. The syntax for this function is as follows:

which(X %in% Y)

The output specifies which parts of X are located in Y. For our purposes we want the rownames(training.object@meta.data) (X) that are located in our doublets.ids (Y). To do this, we will execute the following code:

```{r identify.doublet.rows}
doublet.rows<-which(rownames(training.object@meta.data) %in% doublets.ids)
doublet.rows
```

Notice the output of this code is a series of numbers. These numbers correspond to which rows in our @meta.data slot match the names of the cell IDs in our doublet.ids slot. For instance, the second row of the @meta.data slot should be the same as the second doublet.ids. We can confirm one of them quickly:

```{r confirm.doublet.rows}
rownames(training.object@meta.data)[2]
doublets.ids[2]
```

You can see that those two cell IDs are identical. We have stashed this row information in the doublet.rows variable.

Now that we have identified which rows in our meta.data matches our doublets, we can update our list we created earlier to reflect this. We will want to rename the components of our list that match the rows of the meta data column to instead say "doublet", rather than "singlet". To do that, we execute the following code:

```{r update.doublet.list}
doublets.metadata[doublet.rows]<-"doublets"
doublets.metadata[c(1:3,12:15)]
```

We can see now that our new meta data list contains both doublet and singlet information about all of our cells. The final step is to convert our list into a series of characters and add that information to our Seurat object as new meta data. To do this we will use as.character():

```{r doublet.character}
doublets.metadata<-as.character(doublets.metadata)
doublets.metadata[c(1:3,13:15)]
```

You can see that we now have a series of characters, rather than a list. Remember that Seurat only accepts a list of characters as meta data, not a list. Let us finally add this new meta data information to our Seurat object.

```{r add.doublet.meta.data}
training.object[["doublets"]]<-doublets.metadata
colnames(training.object@meta.data)
unique(training.object@meta.data$doublets)
```

When you add meta data to a Seurat object you will need to specify a name for the new information in the object[["NAME"]]. You can see that we have successfully added our new meta data to the column, and that the information is telling us which cells are singlets or doublets.

We could even visualize where those cells are located in our dimentionality reduction:

```{r visualize.doublets}
DimPlot(training.object,group.by = "doublets")
```

# Adding New Meta Data from Information Already Present in the Meta Data

In the previous scenario, we were adding new meta data to our Seurat object given known cell ids. In this scenario we will be using information already present in the @meta.data slot to generate new meta data. In this example we have analyzed the marker genes from our unbiased clustering and have identified the following information:

1) Cells from cluster A1 represent our microglia
2) Cells from cluster A2 represent our astrocytes
3) Cells from cluster A3 represent our neurons

We have been tasked with updating our cell meta data to reflect this new information. To accomplish this, we will utilize tools we learned in the previous scenario in addition to some new tools.

Just as before, the first task is to create an empty list where we can add in our new meta data:

```{r cell.type.list}
cell.types<-list()
cell.types
```

Next we need to identify which column in our @meta.data slot contains the information about the unbiased clustering labels. We can see that this information is stored in the "paper.ident" column of our @meta.data slot.

```{r paper.ident.slot}
unique(training.object@meta.data$paper.ident)
```

Our next task is to identify which rows of our @meta.data correspond to "A1", "A2", and "A3" respectively. 

To do this, we will use the grep() function. grep() is very similar to the which() function, in that it is used to find a specific string in a given variable. However, grep() allows you to find only part of a pattern, while which() searches for an exact pattern. The syntax for grep() is as follows:

grep(pattern,variable)

This will output which parts of the variable contain the selected pattern. For our purposes, we want to know which rows of training.object@meta.data$paper.ident contain the pattern "A1","A2", and "A3". Let's start with "A1"

```{r pattern.match.a1}
A1.rows<-grep("A1",training.object@meta.data$paper.ident)
A1.rows
```

Notice the output of this code is a series of numbers. Just like in the which() function, these numbers correspond to which rows in our @meta.data$paper.ident are called "A1". Let's confirm one real quick:

```{r confirm.pattern.a1}
training.object@meta.data$paper.ident[3]
```

Now we need to repeat the same procedure for the "A2" and "A3" idents:

```{r pattern.match.a2.a3}
A2.rows<-grep("A2",training.object@meta.data$paper.ident)
A3.rows<-grep("A3",training.object@meta.data$paper.ident)
```

Just like before, we will need to add these cell identities into our new meta data list that we created earlier (cell.types). However, we also need to remember the information that we had from before:

1) Cells from cluster A1 represent our microglia
2) Cells from cluster A2 represent our astrocytes
3) Cells from cluster A3 represent our neurons

So when we name the new components of our list, we don't want to call them A1, A2, A3 again, but instead call them microglia, astrocytes, and neurons. Let's go ahead and try it with just the microglia (A1):

```{r add.a1.to.metadata}
cell.types[A1.rows] <- "microglia"
head(cell.types)
```

Just like in the previous scenario, we are updating the components of our meta data list that we want to rename (cell.types[A1.rows]) and then we state what we want those components to be renamed (<- "microglia").

Notice, however, that some of our components are listed as NULL. The reason why is that we did not give a consecutive list of components to be renamed, but rather only those that matched our A1 cell designation. As such, R automatically filled in the missing values with NULL. Whenever you specify new components of a list, the list will automatically fill in values that you do not specify, and will also expand the list to add in the new values you specify. As an example, let's look at how long our list is:

```{r check.list.length}
length(cell.types)
```

Right now, our list contains 241 components. Let's go ahead and add the A2 and A3 designations to our list, remembering to specify that A2 are astrocytes and A3 are neurons

```{r add.a2.a3}
cell.types[A2.rows] <- "astrocytes"
cell.types[A3.rows] <- "neurons"
length(cell.types)
```

You can see that our list automatically expanded to include the additional components that we added. Let's go ahead and confirm that we have added the correct information to our list:

```{r confirm.list.labels}
cell.types[1:10]
```

As you can see, we have now added information about all the cell types in the correct order. The final step is to add this new information to our meta data object. Remember, that our cell.types is a list, but Seurat only accepts a series of characters for meta data!

```{r add.cell.types.metadata}
cell.types<-as.character(cell.types)
training.object[["cell.types"]]<-cell.types
colnames(training.object@meta.data)
unique(training.object@meta.data$cell.types)
```

We successfully added new information to our meta data from prior information! Let's go ahead and visualize our new information:

```{r visualize.cell.types}
DimPlot(training.object, group.by="cell.types")
```

##Calculating Module (Gene) Scores for pathways of interest

Sometimes you may wish to understand which of the clusters of cells express a particular pathway or set of genes that is relevant to the biology you are looking at. For example, you be studying a cancer cell that has a mutation in the APC complex, and you want to know which populations of cells have upregulated Wnt Signaling.

To address these issues, it is possible to calulate a single value that encompasses the overall level of activity for a larger set of genes. This is performed with the AddModuleScore() command in Seurat. Before we discuss how to run the commands, first it is important to understand the theory and logic behind the calculation, so we can better interpret the results.

The module score calculation is a fairly simple mathematical calculation that compares a given input set of genes (Wnt Signaling) and compares the expression level of those genes to the average expression of 100 randomized gene sets. Importantly, the randomized gene set is generated from genes with similar expression levels as that of the input set. To accomplish this, Seurat first categorizes every gene in our dataset into bins based on their expression level. An example of this is shown below:

```{r fake.histogram, echo=FALSE}
histogram.data<-data.frame("number"=c(1:10),"expression"=c(rev(1:10)),"")

ggplot(histogram.data, aes(x=expression,y=number))+
  geom_bar(stat = "identity")+
  theme_classic()+
  ylab("Number of Genes")+
  xlab("Low Expression > High Expression")
```

As we can see from this example, all of the genes in our dataset are binned into 10 categories, based on their relative expression to each other across the dataset. Notice that the majority of the genes in our datasets will be lowly expressed across all of the cells. In fact, the more realistic plot is substantially more right-skewed. Let's now say that we want to look at the pathway score for a set of genes that we are interested in. Let's call them Gene A, Gene B, and Gene C. If we look at them in our data we find that:

1) Gene A is very highly expressed
2) Gene B is moderately expressed
3) Gene C is lowly expressed

This is where they are located on our histogram example.

```{r fake.histogram.color.genes, echo=FALSE}
histogram.data$gene<-c("Gene A", rep("Other Genes",4), "Gene B", rep("Other Genes",3),"Gene C")

ggplot(histogram.data, aes(x=expression,y=number,fill=gene))+
  geom_bar(stat = "identity")+
  theme_classic()+
  ylab("Number of Genes")+
  xlab("Low Expression > High Expression")+
  scale_fill_manual(values=c("#7fc97f", "#beaed4", "#fdc086", "grey"))+
  theme(axis.ticks = element_blank(), axis.text = element_blank())
```

The next step of the process is to randomly sample 100 genes from each bin that Gene A, Gene B, and Gene C fall into. In this way, our randomized "background" set of genes consists of 300 genes in total, with the same distribution as the genes of interest. The function then calculates the average expression of our three genes of interest and the average expression of the 300 genes of interest on a cell-specific level.

Average(Genes of Interest)
Average(Randomized Genes)

To determine the gene score, it is simply the difference in the two averages:

Module Score = Average(Genes of Interest) - Average(Randomized Genes)

Notice, that this calculation means that: 

1) A score of 0 indicates no difference between your gene set and a randomized gene set
2) A score > 0 indicates upregulation or higher expression of your gene set compared to a randomized gene set
3) A score < 0 indicates downregulation or lower expression of your gene set compared to a randomized gene set

Now for the examples of calculating a gene score in Seurat. First, we need to load in an example set of genes. Execute the following code:

```{r load.example.gene.set}
gene.set <- read.table("gene_set.txt")
```

Let's go ahead and look at the format of our gene list:

```{r head.gene.set}
head(gene.set)
```

We can see that this represents a dataframe with one column and multiple rows representing a different gene in the pathway that we are interested in (total of 135 genes). In order for us to use this gene set in the Seurat AddModuleScore() function, we have to first turn it into a string of characters and then into a list.

```{r convert.gene.set.to.characters}
gene.set <- as.character(gene.set$V1)
gene.set <- list("genes" = gene.set)
```

Now we can go ahead and input this gene set into the Seurat AddModuleScore function. Execute the following code:

```{r add.module.score}
training.object <- AddModuleScore(training.object, features = gene.set, assay = "RNA")
```

##Running Seurat Tutorial - Clustering and Marker Gene Identification

```{}